pex/cache/data.py

```diff
@@ -0,0 +1,434 @@
+
+
+def _iter_key_chunks(
+    items,  # type: Sequence
+    extract_key=None,  # type: Optional[Callable[[_I], _K]]
+):
+    # type: (...) -> Iterator[Tuple[str, Sequence[_K]]]
+
+    # N.B.: Maximum parameter count is 999 in pre-2020 versions of SQLite 3; so we limit
+    # to an even lower chunk size to be safe: https://www.sqlite.org/limits.html
+    chunk_size = 100
+    for index in range(0, len(items), chunk_size):
+        item_chunk = items[index : index + chunk_size]
+        keys = tuple(map(extract_key, item_chunk) if extract_key else item_chunk)
+        placeholders = ", ".join(itertools.repeat("?", len(keys)))
+        yield placeholders, keys
+
+
+def _zipapp_deps(
+    unzip_dirs,  # type: Sequence[UnzipDir]
+    connection=None,  # type: Optional[sqlite3.Connection]
+):
+    # type: (...) -> Iterator[Union[BootstrapDir, UserCodeDir, InstalledWheelDir]]
+
+    with _db_connection(conn=connection) as conn:
+        for placeholders, keys in _iter_key_chunks(unzip_dirs, extract_key=lambda u: u.pex_hash):
+            with closing(
+                conn.execute(
+                    """
+                    SELECT bootstrap_hash, code_hash FROM zipapps WHERE pex_hash IN ({keys})
+                    """.format(
+                        keys=placeholders
+                    ),
+                    keys,
+                )
+            ) as cursor:
+                bootstrap_hash, code_hash = cursor.fetchone()
+                yield BootstrapDir.create(bootstrap_hash)
+                yield UserCodeDir.create(code_hash)
+
+            with closing(
+                conn.execute(
+                    """
+                    SELECT name, install_hash, wheel_hash FROM wheels
+                    JOIN zipapp_deps ON zipapp_deps.wheel_install_hash = wheels.install_hash
+                    JOIN zipapps ON zipapps.pex_hash = zipapp_deps.pex_hash
+                    WHERE zipapps.pex_hash in ({keys})
+                    """.format(
+                        keys=placeholders
+                    ),
+                    keys,
+                )
+            ) as cursor:
+                for name, install_hash, wheel_hash in cursor:
+                    yield InstalledWheelDir.create(
+                        wheel_name=name, install_hash=install_hash, wheel_hash=wheel_hash
+                    )
+
+
+def _venv_deps(
+    venv_dirs,  # type: Sequence[VenvDirs]
+    connection=None,  # type: Optional[sqlite3.Connection]
+):
+    # type: (...) -> Iterator[InstalledWheelDir]
+
+    with _db_connection(conn=connection) as conn:
+        for placeholders, keys in _iter_key_chunks(venv_dirs, extract_key=lambda v: v.short_hash):
+            with closing(
+                conn.execute(
+                    """
+                    SELECT name, install_hash, wheel_hash FROM wheels
+                    JOIN venv_deps ON venv_deps.wheel_install_hash = wheels.install_hash
+                    JOIN venvs ON venvs.short_hash = venv_deps.venv_hash
```

Comment:
There is no need for this join and there is also no need for the venvs table at all any more.